### ISSUE TITLE
RFC: Remove proxy classes.

### ISSUE THREAD

**Issue by DavertMik** [goaop/framework#95]
Hi, according to my previous issue #94 I'd like to propose the idea to remove proxy classes. If its possible. Actually proxy classes are bad as they 
- increase complexity of hierarchy
- you don't know where you are: `$this` points to proxy class.
- it's hard to read code of proxified class (I know they are not designed to read, but)
- file with class definition is 1.5 longer then initial class because of Proxy class.
- and from previous, you can't debug with XDebug because you are pointed to the wrong line.

A simplified approach I'd suggest to reduce the level of dark magic :)

Basic class:

``` php
<?php
class User {
     function register($email, $password)
    { 
        $user = new User
        $user->setEmail($email);
        $user->setPassword($password);
        $user->save();
    }
}
?>
```

Transforms to:

``` php
<?php
class User implements JoinPoints {
     function register($email, $password)
    { $__res = joinpoint_before('method:register', $this, array($email, $password))); if ($__res instanceof ExecutionTerminated) return $__res->getResult();  
        $user = new User
        $user->setEmail($email);
        $user->setPassword($password);
        $user->save();
    joinpoint_after('method:register', $this, array($email, $password)); }
}
?>
```

Thus, enhanced class is just like the original one. But joinpoints are set and they work as we expect them. We can set before/after/around aspects. joinpoint methods were added to the exactly the same lines a braces are. Thus, we didn't change the line positions in file.

**UPS** Looks like in this case we can't intercept method that returns something. Thus we will need to hook over the **return** statement. Which I'm not sure is actually possible.

**Comment by lisachenko** [comment]
Thank you for the RFC. I definitely don't want to change the original code by weaving an advices into the concrete method because it's a wrong way. Look at the cemetery of AOP frameworks (phpAspect, apdt, etc) that used such transformation techniques and you will get an idea why it's bad.

I prefer current object-oriented approach with invocations as it's more straightforward. Of course, you are right that there are some issues with current approach, but it's the best implementation for performance/caching reasons.

One thing that I want to implement is to split modified original class and proxy to the separate files. This can give more control with XDebug and still can give an ability to put a breakpoint in the original code. This is very important requirement for my framework.

Feedback is appreciated :)

**Comment by DavertMik** [comment]
Thanks. Yes, implementation with "changing the code itself" is not much simpler and has its own drawbacks.
Great to know, that you have ideas how to improve XDebug execution. 
I'm using it quite often to get a clue why must test did fail.

**Comment by lisachenko** [comment]
Ok, I'll try to do my best for proxy separation, but I'm too busy now, so you shouldn't expect this feature soon.

BTW, is it so uncomfortable for you to debug an original class now? You can run the code in production mode with XDebug to see how it works, so there won't be any magic with XDebug at all, but you should put your breakpoints in the cache file instead of original file.

**Comment by DavertMik** [comment]
I ended with creating my own transformer. 
https://github.com/Codeception/AspectMock/blob/master/src/AspectMock/Intercept/BeforeMockTransformer.php

Thanks for this really flexible architecture.
I realized that I don't need the `after` calls, and with adding just a one line into beginning of each method looks just fine. Lots of problems with proxies solved, debug works as a charm. I'm pretty happy :)

Btw, there should be a way to completely ignore some classes in `AopComposerLoader`. Currently only Go files are ignored, but I'd like to add AspectMock files into ignore list as well. Will it be possible?

**Comment by lisachenko** [comment]
> Btw, there should be a way to completely ignore some classes in AopComposerLoader. Currently only Go files are ignored, but I'd like to add AspectMock files into ignore list as well. Will it be possible?

Your idea is good, this can be implemented. Easiest way to do this is to add static method for AopComposerLoader, but this is ugly. Better way is to use configuration (eg. composer `extra` field, kernel configuration, etc)

**Comment by lisachenko** [comment]
> I ended with creating my own transformer.
> https://github.com/Codeception/AspectMock/blob/master/src/AspectMock/Intercept/BeforeMockTransformer.php
> 
> Thanks for this really flexible architecture.

Your are always welcome :)

**Comment by lisachenko** [comment]
Implemented proxy separation in #102, however there is a chance of bug with `ReflectionClass->getFileName()` for proxies (#108).
